### PR TITLE
Adding Nest Thermostat support - initial set

### PR DESCRIPTION
**Pull request by sanjaiganesh**
Pending
- async support
- Implement rest of the methods in the schema
- adopt an unit test framework

### CODE REVIEW

File: org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/README.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
+access_token :
```
Comment (tjaffri) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/README.md]: Is the first part of this README required? lines 1-11? Not sure if you want to check in your access token like that either (somebody will be able to use it to mess with your thermostat)

File: org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        var postData;
+        if (scale == "c"){
+            postData = { name : value };
+        }
+        else {
+            postData = { name : value };
+        }
+        postData = JSON.stringify(postData);
+
+        var options = {
```
Comment (tjaffri) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js]: Are you going to refactor out a opent2t-translator-helper-nest like @rlingineni did with his opent2t-translator-helper-wink in his pull request?

**Comment by jasongin** [comment]
@sanjaiganesh Since you are a MS employee you actually do not need to sign a CLA. Please follow the [instructions to link your github account to the Microsoft org](https://opensourcehub.microsoft.com/articles/how-to-join-microsoft-github-org-self-service).

File: org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml [hunk 3]
```diff
@@ -10,10 +10,20 @@
...
     <!-- Get the current ambient temperature -->
-    <method name="getCurrentTemperature">
+    <method name="getAmbientTemperature">
+      <!-- Temperature scale:  'f' for Fahrenheit. 'c' for celsius -->
+      <arg name="scale" type="s" direction="in" />
       <arg name="temp" type="d" direction="out" />
     </method>
 
+    <!-- Set the current ambient temperature -->
+    <method name="setTargetTemperature">
```
Comment (jasongin) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml]: There are already methods for this below, they are called setHeatingSetpoint and setCoolingSetpoint. Is there a reason they don't work for Nest?

File: org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+                        return;
+                    }
+                } else {
+                    var data = JSON.parse(body);
+                    if ((scale == "c") && !!data.target_temperature_c) {
+                        callback(data.target_temperature_c);
+                    }
+                    else if ((scale == "f") && !!data.target_temperature_f) {
+                        callback(data.target_temperature_f);
+                    }
```
Comment (jasongin) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js]: If the requested unit doesn't match the unit reported by the thermostat, then you need to either do the conversion or report an error. As it is here, the async call never returns in that case.

File: org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+                        return;
+                    }
+                } else {
+                    var data = JSON.parse(body);
+                    callback(data[name]);
+                    }
+            });
+
+            getRes.on('error', function(e) {
+                callback('error');
```
Comment (jasongin) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js]: I know you mentioned this in the summary comments... you should convert all these methods to return promises instead of using callbacks. Then the error handling will be cleaner too: you can reject the promise with the error object.

File: org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+                if (typeof (props.id) !== 'undefined') {
+                    deviceId = props.id;
+                } else {
+                    console.log('props.id is undefined.');
+                }
+            } else {
+                console.log('props is undefined.');
+            }
+        } else {
+            console.log('device is undefined.');
```
Comment (jasongin) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js]: Invalid arguments should result in thrown Errors instead of just logging. (I know you copy-pasted this part from other translators.) This can be made much cleaner using [a validation helper function like this](https://github.com/openT2T/translators/blob/master/org.OpenT2T.Sample.SuperPopular.Lamp/Philips%20Hue%20Light%20Bulb/js-promise/thingTranslator.js#L33).

File: org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml (hunk 3 above)
Comment (sanjaiganesh) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml]: I shall add getTargetTemperature as well

File: org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/README.md (hunk 1 above)
Comment (sanjaiganesh) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/README.md]: Oops.. this must be removed.. I had saved this for testing purposes.

---
In reply to: [68174201](https://github.com/openT2T/translators/pull/22#discussion_r68174201) [](ancestors = 68174201)

File: org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml (hunk 3 above)
Comment (sanjaiganesh) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml]: My understanding was, they are min/max values for heating/cooling.. 

Tariq, can you pls clarify?

The way I mapped..

set/getHeatingSetpoint
    "target_temperature_high_c": 26,
    "target_temperature_high_f": 79,

set/getCoolingSetpoint
    "target_temperature_low_c": 19,
    "target_temperature_low_f": 66,

get/setTargetTemperature (new method that is added). I need to add 'get' as well.
   "target_temperature_c": 24,
    "target_temperature_f": 75,

getAmbientTemperature (formerly getCurrentTemperature)
    "ambient_temperature_c": 11,
    "ambient_temperature_f": 52,




---
In reply to: [68280848](https://github.com/openT2T/translators/pull/22#discussion_r68280848) [](ancestors = 68280848)

File: org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js (hunk 2 above)
Comment (sanjaiganesh) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js]: I shall check it out.

---
In reply to: [68174265](https://github.com/openT2T/translators/pull/22#discussion_r68174265) [](ancestors = 68174265)

File: org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml [hunk 7]
```diff
@@ -10,10 +10,20 @@
     <method name="turnOff"/>
     
     <!-- Get the current ambient temperature -->
-    <method name="getCurrentTemperature">
+    <method name="getAmbientTemperature">
+      <!-- Temperature scale:  'f' for Fahrenheit. 'c' for celsius -->
+      <arg name="scale" type="s" direction="in" />
```
Comment (jasongin) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml]: I'd prefer the term "unit" over "scale". But if we're adding a unit parameter then other thermostat translators (currently only Wink) will need to be updated to check the unit and possible do a conversion.

File: org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml [hunk 8]
```diff
@@ -10,10 +10,20 @@
     <method name="turnOff"/>
     
     <!-- Get the current ambient temperature -->
-    <method name="getCurrentTemperature">
+    <method name="getAmbientTemperature">
```
Comment (jasongin) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml]: Changing the schema breaks existing translators that implement it. So you'd need to rename the corresponding method in the Wink thermostat translator, and update the test too.

File: org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+
+    setTemperatureProperty: function(name, scale, value, callback) {
+
+        var postData;
+        if (scale == "c"){
+            postData = { name : value };
+        }
+        else {
+            postData = { name : value };
+        }
```
Comment (sanjaiganesh) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js]: if/else not required. Will replace it with just 'postData = { name : value };'

File: org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml (hunk 7 above)
Comment (sanjaiganesh) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml]: From the documentation and in general, i saw that 'temperature scale' usage.. so I preferred it over 'units' .

Taqi /Ravi ?

---
In reply to: [68283058](https://github.com/openT2T/translators/pull/22#discussion_r68283058) [](ancestors = 68283058)

File: org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml (hunk 8 above)
Comment (sanjaiganesh) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml]: Wink is not using using thermostat schema. Nest is the first thermostat being added here. 
Wink is using org.OpenT2T.Sample.SuperPopular.TemperatureSensor schema. Am I missing anything ?

---
In reply to: [68283197](https://github.com/openT2T/translators/pull/22#discussion_r68283197) [](ancestors = 68283197)

File: org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js (hunk 6 above)
Comment (sanjaiganesh) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js]: Sure. 

---
In reply to: [68282541](https://github.com/openT2T/translators/pull/22#discussion_r68282541) [](ancestors = 68282541)

File: org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js (hunk 5 above)
Comment (sanjaiganesh) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js]: sure. 

---
In reply to: [68281720](https://github.com/openT2T/translators/pull/22#discussion_r68281720) [](ancestors = 68281720)

File: org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml (hunk 3 above)
Comment (jasongin) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml]: I don't see a translator implementation for the get/set Heating/Cooling setpoint methods, so I thought you were adding setTargetTemperature instead of those. But apparently it is a 3rd (middle?) target temperature that Nest defines in addition to the usual high/low targets? How does that work?

---
In reply to: [68282643](https://github.com/openT2T/translators/pull/22#discussion_r68282643) [](ancestors = 68282643)

**Comment by sanjaiganesh** [comment]
Ah ok. Sure.

---
In reply to: [228124748](https://github.com/openT2T/translators/pull/22#issuecomment-228124748) [](ancestors = 228124748)

File: org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml (hunk 8 above)
Comment (jasongin) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml]: Oh, the Wink thermostat translator is not merged yet, but it's about to be since I just signed off. [https://github.com/openT2T/translators/pull/18](https://github.com/openT2T/translators/pull/18)

---
In reply to: [68284485](https://github.com/openT2T/translators/pull/22#discussion_r68284485) [](ancestors = 68284485,68283197)

File: org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml (hunk 7 above)
Comment (rlingineni) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml]: I like the sound of 'Unit' better too, but this is definitely going to break the current translators. The Wink thermostat is currently in a pull request, and it will be merged soon. I think one good practice is to make any schema changes in a separate PR because they can potentially break existing code. 

Personally I feel like Celsius (for now) should be the preferred unit and then we don't worry about error handling and parameters and we make it clear in the documentation. The conversion I feel should be handled client-side.

Comment (jasongin) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml]: I don't feel strongly about the naming right now. At some point later we'll probably have a broader discussion about how to handle units in the schemas and translators, and we can update everything to be consistent then.

However the Wink thermostat will need to be updated if this parameter is added.

---
In reply to: [68283792](https://github.com/openT2T/translators/pull/22#discussion_r68283792) [](ancestors = 68283792,68283058)

File: org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml (hunk 8 above)
Comment (rlingineni) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml]: Yeah so it's currently in PR at #18 waiting to be merged, we just fixed a bunch of errors. So it will be merged soon

File: org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml (hunk 3 above)
Comment (rlingineni) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml]: I think it might be in reference to "Auto" Mode. I feel that we could still use setHigh and setLow. Wink doesn't support setTarget from my understanding. It wouldn't be good to have that.

File: org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml (hunk 8 above)
Comment (sanjaiganesh) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml]: Got it. I shall merge and fix it as well.

---
In reply to: [68286096](https://github.com/openT2T/translators/pull/22#discussion_r68286096) [](ancestors = 68286096)

Comment (sanjaiganesh) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml]: But can you accommodate reading both C/F units in it and when I merge I will hook the 'scale/unit' args to it.

---
In reply to: [68288119](https://github.com/openT2T/translators/pull/22#discussion_r68288119) [](ancestors = 68288119,68286096)

Comment (rlingineni) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml]: That shouldn't be a problem, but ideally your schema should be merged first before we can make that change. Also, in the Wink Thermostat, since Wink only accepts Celsius, we would need to add a helper method called `converter`

Comment (tjaffri) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml]: I merged #18 since it has been sitting for a while and it had a cool :shipit: cat :). I suggest if any changes need to be made there we add the stub implementations in this PR, then Ravi can update the stubs with full implementations later in a later PR. Cool?

Comment (sanjaiganesh) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml]: sure. sounds good.

---
In reply to: [68289826](https://github.com/openT2T/translators/pull/22#discussion_r68289826) [](ancestors = 68289826)

Comment (rlingineni) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml]: Makes sense. Is it possible to make a separate PR for any schema changes though, so it's really obvious or could you update the name of this PR to "Modified Schema and Added Nest Thermostat"

Comment (jasongin) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/org.OpenT2T.Sample.SuperPopular.Thermostat.xml]: FYI it's actually a [squirrel](https://www.quora.com/GitHub-What-is-the-significance-of-the-Ship-It-squirrel).

**Comment by sanjaiganesh** [comment]
Did this and linked the accounts. Thanks.

---
In reply to: [228133392](https://github.com/openT2T/translators/pull/22#issuecomment-228133392) [](ancestors = 228133392,228124748)

**Comment by sanjaiganesh** [comment]
#Closed

---
In reply to: [227920635](https://github.com/openT2T/translators/pull/22#issuecomment-227920635) [](ancestors = 227920635)

File: org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js (hunk 4 above)
Comment (sanjaiganesh) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js]: I am removing ability to specify units, to keep things simple. No changes required for wink thermostat implementation. 

---
In reply to: [68281410](https://github.com/openT2T/translators/pull/22#discussion_r68281410) [](ancestors = 68281410)

**Comment by sanjaiganesh** [comment]
@jasongin @tjaffri @rlingineni 
Address PR comments. added helper. please take a look.

**Comment by axis-tip** [comment]
spoke to @jasongin and he will review the latest. Stay tuned.

File: org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        console.log('isTurnedOn called.');
+        return helper.getProperty('hvac_state').then(state => {
+            console.log("state: "+ state);
+            return state != 'off';
+        }).catch(error => {
+            throw error;
+        });
+    },
+    
+    // Default to Heating. THe call must be followed by setMode(heating/cooling), as desired.
```
Comment (jasongin) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js]: I think "heat-cool" would be a more natural default.

However, if the hvac is already on, this turnOn command might change the mode, which seems like strange behavior. So should you check whether it is off first?

File: Helpers/nest-helper/js/index.js [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        var httpsRequest = https.get(options, function(getRes) {
+            var body = '';
+            getRes.setEncoding('utf8');
+            getRes.on('data', function(data) {
+                body += data;
+            });
+
+            getRes.on('end', function() {
+                if (getRes.statusCode != 200) {
+                        deferred.reject("Error Code:" + getRes.statusCode);
```
Comment (jasongin) [inline Helpers/nest-helper/js/index.js]: The value passed to reject() should be an instance of an Error object. Also on line 81.

File: Helpers/nest-helper/js/index.js [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+        });
+        httpsRequest.end();
+
+        return deferred.promise; // return the promise
+    },
+    
+    setProperty : function(name, value) {
+        var deferred = q.defer();   // Take a deferral
+
+        var postData = JSON.stringify({name : value});
```
Comment (jasongin) [inline Helpers/nest-helper/js/index.js]: This results in an object with a property literally named "name", which is not what you intended.

**Comment by jasongin** [comment]
> 

You don't need these lines. The translators only did this for methods which were directly exposed in the translator schema, which doesn't apply here.

---
Refers to: Helpers/nest-helper/js/index.js:102 in 5069258. [](commit_id = 50692589416d07c52bbf52c6fab02d6b109f6561, deletion_comment = False)

File: Helpers/nest-helper/js/index.js [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+            getRes.on('data', (data) => {
+                body += data;
+            });
+
+            getRes.on('end', () => {
+                if (getRes.statusCode != 200) {
+                    deferred.reject("Error Code:" + getRes.statusCode);
+                } else {
+                    var data = JSON.parse(body);
+                    deferred.resolve(target_temperature_c);
```
Comment (jasongin) [inline Helpers/nest-helper/js/index.js]: Undefined variable: target_temperature_c.
I don't think you even need to parse the body or return any value: if the status is 200 then just deferred.resolve().

File: Helpers/nest-helper/js/index.js (hunk 11 above)
Comment (sanjaiganesh) [inline Helpers/nest-helper/js/index.js]: sure. 

---
In reply to: [68797598](https://github.com/openT2T/translators/pull/22#discussion_r68797598) [](ancestors = 68797598)

**Comment by sanjaiganesh** [comment]
> 

sure.

---
In reply to: [229113564](https://github.com/openT2T/translators/pull/22#issuecomment-229113564) [](ancestors = 229113564)

---
Refers to: Helpers/nest-helper/js/index.js:102 in 5069258. [](commit_id = 50692589416d07c52bbf52c6fab02d6b109f6561, deletion_comment = False)

File: Helpers/nest-helper/js/index.js (hunk 13 above)
Comment (sanjaiganesh) [inline Helpers/nest-helper/js/index.js]: ah . thanks for the catch.

---
In reply to: [68798887](https://github.com/openT2T/translators/pull/22#discussion_r68798887) [](ancestors = 68798887)

File: Helpers/nest-helper/js/index.js (hunk 12 above)
Comment (sanjaiganesh) [inline Helpers/nest-helper/js/index.js]: great catch. thanks. I have issues in verifying 'PUT'.. it gets struck.. so couldn't validate properly. I now see that it works once in 10-15 attempts. I am assuming it could an issue with redirects (testing with Nest simulator).. I may be wrong..

---
In reply to: [68798191](https://github.com/openT2T/translators/pull/22#discussion_r68798191) [](ancestors = 68798191)

File: Helpers/nest-helper/js/index.js (hunk 11 above)
Comment (jasongin) [inline Helpers/nest-helper/js/index.js]: You still need the next two lines of code to be inside the else{} block (that was removed). Because deferred.reject() doesn't directly throw an error, this will reject() and then resolve() immediately afterward... I don't know how that will behave.

---
In reply to: [68817345](https://github.com/openT2T/translators/pull/22#discussion_r68817345) [](ancestors = 68817345,68797598)

File: Helpers/nest-helper/js/index.js (hunk 13 above)
Comment (jasongin) [inline Helpers/nest-helper/js/index.js]: Again, the resolve() still needs to be in an else{} block, because the Error is not thrown.

---
In reply to: [68837920](https://github.com/openT2T/translators/pull/22#discussion_r68837920) [](ancestors = 68837920,68798887)

File: org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js (hunk 10 above)
Comment (jasongin) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js]: Still waiting for a resolution here.

---
In reply to: [68792856](https://github.com/openT2T/translators/pull/22#discussion_r68792856) [](ancestors = 68792856)

File: org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+    },
+    
+    isTurnedOn : function() {
+        console.log('isTurnedOn called.');
+        return helper.getProperty('hvac_mode').then(state => {
+            console.log("state: "+ state);
+            return state != 'off';
+        }).catch(error => {
+            throw error;
+        });
```
Comment (jasongin) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js]: This is redundant if all it's going to do is catch and throw again. You might have seen other code that did this but also logged the error.

**Comment by sanjaiganesh** [comment]
I think the change can be merged. Addressed all the comments.

**Comment by jasongin** [comment]
@sanjaiganesh There are still 3 issues that are not addressed. Unfortunately those comments are now hidden by the github web UI because they are against older commits, but they are still valid comments. They are easier to see in CodeFlow: look for issues opened by me that are still Active. (I have Closed all the issues that you resolved.)

Let me see if I can link directly to them...
https://github.com/openT2T/translators/pull/22#discussion_r68792856
https://github.com/openT2T/translators/pull/22#discussion_r68851382
https://github.com/openT2T/translators/pull/22#discussion_r68851587

**Comment by sanjaiganesh** [comment]
Ah ok. sure thing.. It is a bit difficult to track GIT CR comments, as I can't close comments created by others.

File: org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js (hunk 10 above)
Comment (sanjaiganesh) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js]: sure. Now setting to heat-cool only if it is off. otherwise turnOn is no-op. 

---
In reply to: [68851734](https://github.com/openT2T/translators/pull/22#discussion_r68851734) [](ancestors = 68851734,68792856)

File: Helpers/nest-helper/js/index.js (hunk 11 above)
Comment (sanjaiganesh) [inline Helpers/nest-helper/js/index.js]: oh.. i assumed that it will throw, which is wrong. will fix that.

---
In reply to: [68851382](https://github.com/openT2T/translators/pull/22#discussion_r68851382) [](ancestors = 68851382,68817345,68797598)

File: Helpers/nest-helper/js/index.js (hunk 13 above)
Comment (sanjaiganesh) [inline Helpers/nest-helper/js/index.js]: sure. 

---
In reply to: [68851587](https://github.com/openT2T/translators/pull/22#discussion_r68851587) [](ancestors = 68851587,68837920,68798887)

File: org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+    
+    // Default to Heating. THe call must be followed by setMode(heating/cooling), as desired.
+    turnOn : function() {
+        console.log('turnOn called.');
+        return getMode().then( mode => {
+            if(mode == 'off')
+            {
+                return helper.setProperty({'hvac_mode' : 'heat-cool'});
+            }
+        });
```
Comment (jasongin) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js]: This function still needs to return a promise even when the mode is not off. So add an else block that returns a resolved promise.

Comment (sanjaiganesh) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js]: Sure. Actually I had the same doubt.

Comment (sanjaiganesh) [inline org.OpenT2T.Sample.SuperPopular.Thermostat/Nest Thermostat/js/thingTranslator.js]: Returning an empty promise, as setProperty does. 

---
In reply to: [69182757](https://github.com/openT2T/translators/pull/22#discussion_r69182757) [](ancestors = 69182757)

**Comment by jasongin** [comment]
:shipit:

**Comment by sanjaiganesh** [comment]
thanks. :)

**Comment by sanjaiganesh** [comment]
Thanks for the detailed review.
